File: src/api/modules/logs.ts
import { invoke } from '@tauri-apps/api/core'
import type { ApiResponse, AppLog, PlayHistory, LyricSource } from './types'

// 日志
export async function addLog(level: string, message: string, target?: string): Promise<void> {
  const response = await invoke<ApiResponse<void>>('add_log', { level, message, target })
  if (!response.success) throw new Error(response.error)
}

export async function getLogs(level?: string, limit?: number): Promise<AppLog[]> {
  const response = await invoke<ApiResponse<AppLog[]>>('get_logs', { level, limit })
  if (!response.success) throw new Error(response.error)
  return response.data || []
}

export async function getErrorLogs(): Promise<AppLog[]> {
  const response = await invoke<ApiResponse<AppLog[]>>('get_error_logs')
  if (!response.success) throw new Error(response.error)
  return response.data || []
}

export async function clearLogs(): Promise<number> {
  const response = await invoke<ApiResponse<number>>('clear_logs')
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

export async function getLogCount(): Promise<number> {
  const response = await invoke<ApiResponse<number>>('get_log_count')
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

export async function copyLogsToClipboard(): Promise<string> {
  const response = await invoke<ApiResponse<string>>('copy_logs_to_clipboard')
  if (!response.success) throw new Error(response.error)
  return response.data || ''
}

// 播放历史
export async function addPlayHistory(path: string, duration: number, completed: boolean): Promise<void> {
  const response = await invoke<ApiResponse<void>>('add_play_history', { path, duration, completed })
  if (!response.success) throw new Error(response.error)
}

export async function getPlayHistory(limit?: number): Promise<PlayHistory[]> {
  const response = await invoke<ApiResponse<PlayHistory[]>>('get_play_history', { limit })
  if (!response.success) throw new Error(response.error)
  return response.data || []
}

export async function clearPlayHistory(): Promise<void> {
  const response = await invoke<ApiResponse<void>>('clear_play_history')
  if (!response.success) throw new Error(response.error)
}

export async function getPlayCounts(): Promise<Record<string, number>> {
  const response = await invoke<ApiResponse<Array<[string, number]>>>('get_play_counts')
  if (!response.success) throw new Error(response.error)
  return Object.fromEntries(response.data || [])
}

export async function getSongPlayCount(path: string): Promise<number> {
  const response = await invoke<ApiResponse<number>>('get_song_play_count', { path })
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

// 歌词
export async function getLyrics(path: string): Promise<LyricSource | null> {
  const response = await invoke<ApiResponse<LyricSource | null>>('get_lyrics', { path })
  if (!response.success) throw new Error(response.error)
  return response.data ?? null
}

File: src-tauri/src/commands/logs.rs
use tauri::State;

use crate::database::Database;
use crate::database::AppLog;
use crate::database::PlayHistory;
use super::common::ApiResponse;

#[tauri::command]
pub async fn add_log(
    db: State<'_, Database>,
    level: String,
    message: String,
    target: Option<String>,
) -> Result<ApiResponse<()>, String> {
    match db.add_log(&level, &message, target.as_deref()).await {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_logs(
    db: State<'_, Database>,
    level: Option<String>,
    limit: Option<i64>,
) -> Result<ApiResponse<Vec<AppLog>>, String> {
    match db.get_logs(level.as_deref(), limit).await {
        Ok(logs) => Ok(ApiResponse::ok(logs)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_error_logs(
    db: State<'_, Database>,
) -> Result<ApiResponse<Vec<AppLog>>, String> {
    match db.get_error_logs().await {
        Ok(logs) => Ok(ApiResponse::ok(logs)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn clear_logs(db: State<'_, Database>) -> Result<ApiResponse<usize>, String> {
    match db.clear_logs().await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_log_count(db: State<'_, Database>) -> Result<ApiResponse<i64>, String> {
    match db.get_log_count().await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn copy_logs_to_clipboard(
    db: State<'_, Database>,
) -> Result<ApiResponse<String>, String> {
    use std::fmt::Write;
    
    let logs = match db.get_logs(None, Some(100)).await {
        Ok(logs) => logs,
        Err(e) => return Ok(ApiResponse::err(e.to_string())),
    };

    let mut output = String::new();
    let _ = writeln!(&mut output, "=== JlocalMusic 日志 ===");
    let _ = writeln!(&mut output, "应用版本: v{}", env!("CARGO_PKG_VERSION"));
    let _ = writeln!(&mut output, "平台: {}", std::env::consts::OS);
    let _ = writeln!(&mut output, "架构: {}", std::env::consts::ARCH);
    let _ = writeln!(&mut output, "===================");
    let _ = writeln!(&mut output);

    for log in logs {
        let _ = writeln!(
            &mut output,
            "[{}] [{}] {}",
            log.created_at.format("%Y-%m-%d %H:%M:%S"),
            log.level,
            log.message
        );
    }

    Ok(ApiResponse::ok(output))
}

#[tauri::command]
pub async fn add_play_history(
    db: State<'_, Database>,
    path: String,
    duration: i64,
    completed: bool,
) -> Result<ApiResponse<()>, String> {
    match db.add_play_history(&path, duration, completed).await {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_play_history(
    db: State<'_, Database>,
    limit: Option<i64>,
) -> Result<ApiResponse<Vec<PlayHistory>>, String> {
    match db.get_play_history(limit).await {
        Ok(history) => Ok(ApiResponse::ok(history)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn clear_play_history(
    db: State<'_, Database>,
) -> Result<ApiResponse<()>, String> {
    match db.clear_play_history().await {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_play_counts(
    db: State<'_, Database>,
) -> Result<ApiResponse<Vec<(String, i64)>>, String> {
    match db.get_play_counts().await {
        Ok(counts) => Ok(ApiResponse::ok(counts)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_song_play_count(
    db: State<'_, Database>,
    path: String,
) -> Result<ApiResponse<i64>, String> {
    match db.get_song_play_count(&path).await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn cleanup_nonexistent_songs(
    db: State<'_, Database>,
    base_folder: String,
) -> Result<ApiResponse<usize>, String> {
    match db.cleanup_nonexistent_songs(&base_folder).await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}
